Sources/Build/BuildDelegate.swift

```diff
@@ -0,0 +1,223 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+#if canImport(llbuildSwift)
+
+import Basic
+import llbuildSwift
+
+/// Diagnostic error when a llbuild command encounters an error.
+struct LLBuildCommandErrorDiagnostic: DiagnosticData {
+    static let id = DiagnosticID(
+        type: LLBuildCommandErrorDiagnostic.self,
+        name: "org.swift.diags.llbuild-command-error",
+        defaultBehavior: .error,
+        description: { $0 <<< { $0.message } }
+    )
+
+    let message: String
+}
+
+/// Diagnostic warning when a llbuild command encounters a warning.
+struct LLBuildCommandWarningDiagnostic: DiagnosticData {
+    static let id = DiagnosticID(
+        type: LLBuildCommandErrorDiagnostic.self,
+        name: "org.swift.diags.llbuild-command-warning",
+        defaultBehavior: .warning,
+        description: { $0 <<< { $0.message } }
+    )
+
+    let message: String
+}
+
+/// Diagnostic error when llbuild detects a cycle.
+struct LLBuildCycleErrorDiagnostic: DiagnosticData {
+    static let id = DiagnosticID(
+        type: LLBuildCycleErrorDiagnostic.self,
+        name: "org.swift.diags.llbuild-cycle",
+        defaultBehavior: .error,
+        description: {
+            $0 <<< "build cycle detected: "
+            // $0 <<< { $0.rules.map({ $0.description }).joined(separator: ", ") }
+        }
+    )
+
+    let rules: [BuildKey]
+}
+
+public struct BuildDelegate: BuildSystemDelegate {
+	private let diagnostics: DiagnosticsEngine
+	private let outputStream: OutputByteStream
+
+	public init(diagnostics: DiagnosticsEngine, outputStream: OutputByteStream = stdoutStream) {
+		self.diagnostics = diagnostics
+		self.outputStream = outputStream
+	}
+
+    /// The FileSystem to use, if any.
+    ///
+    /// This is currently very limited.
+    public var fs: llbuildSwift.FileSystem? {
+        return nil
+    }
+
+    /// Called in response to requests for new tools.
+    ///
+    /// The client should return an appropriate tool implementation if recognized.
+    public func lookupTool(_ name: String) -> Tool? {
+        return nil
+    }
+    
+    /// Called to report any form of command failure.
+    ///
+    /// This can may be called to report the failure of a command which has
+    /// executed, but may also be used to report the inability of a command to
+    /// run. It is expected to be used by the client in making decisions with
+    /// regard to cancelling the build.
+    public func hadCommandFailure() {
```

Comment:
We probably want to cancel on response to this? That matches what `swift-build-tool` currently does IIRC

Reply:
Done